xdsl/ir.py

```diff
@@ -274,7 +277,26 @@ def irdl_definition(cls) -> ParamAttrDef:
 
 
 @dataclass
-class Operation:
+class IRNode(object):
+    pass
+
+    def is_ancestor(cls: IRNode, op: IRNode) -> bool:
+        "Returns true if the IRNode is an ancestor of another IRNode."
+        if op is cls:
+            return True
+        if op.parent is None:
+            return False
+        return cls.is_ancestor(op.parent)
+
+    def get_toplevel_object(cls: IRNode) -> IRNode:
+        """Get the operation, block, or region ancestor that has no parents."""
+        if cls.parent is None:
+            return cls
+        return cls.parent.get_toplevel_object()
+
```

Comment:
Can we add `parent_op`, `parent_block`, and `parent_region` here? Essentially, this would be the nearest ancestor that has the right type, so that should work?

Reply:
Note: This change will inhibit us from moving the IRNode to its own file, as we would have a circular dependency (IRNode needs Op/Block/Region for type hints, which themselves inherit from IRNode).

I think if we really want to design our hierarchy in a clean way, this might be a bad idea.

Reply:
I think this hierarchy refactor can be done in some next PR. Thoughts?

Reply:
Yes, we can always think of that in a next PR!
And if the plan is to move `IRNode` outside this file, then yes we can forget it!